jerry-core/ecma/builtin-objects/ecma-builtin-string-prototype.c

```diff
@@ -1657,286 +1528,346 @@ ecma_builtin_string_prototype_object_split (ecma_value_t this_arg, /**< this arg
+            ECMA_FINALIZE (array_length_val);
+            ecma_deref_ecma_string (magic_length_str_p);
+            ecma_deref_ecma_string (array_length_str_p);
+            ecma_deref_ecma_string (substr_str_p);
+          } /* if (!ecma_is_value_null (match_result)) */
 
-            /* 15. */
-            ecma_string_t *array_length_string_p = ecma_new_ecma_string_from_uint32 (new_array_length);
+          ecma_free_value (match_result);
 
-            ecma_value_t put_comp = ecma_builtin_helper_def_prop (new_array_p,
-                                                                  array_length_string_p,
-                                                                  ecma_make_string_value (substr_str_p),
-                                                                  true,
-                                                                  true,
-                                                                  true,
-                                                                  false);
+        } /* while (curr_pos < string_length && !should_return && ecma_is_value_empty (ret_value)) */
 
-            JERRY_ASSERT (ecma_is_value_true (put_comp));
+        if (!should_return && !separator_is_empty && ecma_is_value_empty (ret_value))
+        {
+          /* 14. */
+          ecma_string_t *substr_str_p;
+          substr_str_p = ecma_string_substr (this_to_string_p,
+                                             start_pos,
+                                             string_length);
+
+          /* 15. */
+          ecma_string_t *array_length_string_p = ecma_new_ecma_string_from_uint32 (new_array_length);
+
+          ecma_value_t put_comp = ecma_builtin_helper_def_prop (new_array_p,
+                                                                array_length_string_p,
+                                                                ecma_make_string_value (substr_str_p),
+                                                                true,
+                                                                true,
+                                                                true,
+                                                                false);
+
+          JERRY_ASSERT (ecma_is_value_true (put_comp));
+
+          ecma_deref_ecma_string (array_length_string_p);
+          ecma_deref_ecma_string (substr_str_p);
+        }
+      } /* if (string_length != 0) || !ecma_is_value_empty (ret_value) */
 
-            ecma_deref_ecma_string (array_length_string_p);
-            ecma_deref_ecma_string (substr_str_p);
-          }
-        } /* if (string_length != 0) || !ecma_is_value_empty (ret_value) */
+      ecma_free_value (separator);
 
-        ecma_free_value (separator);
-      } /* if (!ecma_is_value_undefined (arg1)) */
-    } /* if (limit != 0) */
-  } /* if (ecma_is_value_empty (ret_value)) */
+    } /* if (!ecma_is_value_undefined (arg1)) */
+  } /* (ecma_is_value_empty (ret_value) && limit != 0) */
```

Comment:
My problem with them is to keep the comments sync with the condition. Some of them are not matching already. :(